server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java

```diff
@@ -0,0 +1,244 @@
+      dbClient.componentDao().insert(componentContext.dbSession, component);
+      return component;
+    } else {
+      ComponentDto component = createComponent(reportComponent, componentKey, existingComponent.uuid(), moduleParent);
+      if (updateComponent(existingComponent, component)) {
+        dbClient.componentDao().update(componentContext.dbSession, existingComponent);
+      }
+      return existingComponent;
+    }
+  }
+
+  private ComponentDto createComponent(BatchReport.Component reportComponent, String componentKey, String uuid, @Nullable ComponentDto parentModule) {
+    ComponentDto component = new ComponentDto();
+    component.setUuid(uuid);
+    component.setKey(componentKey);
+    component.setDeprecatedKey(componentKey);
+    component.setEnabled(true);
+    component.setScope(getScope(reportComponent));
+    component.setQualifier(getQualifier(reportComponent));
+    component.setName(getFileName(reportComponent));
+
+    if (isProjectOrModule(reportComponent)) {
+      component.setLongName(component.name());
+      if (reportComponent.hasDescription()) {
+        component.setDescription(reportComponent.getDescription());
+      }
+    } else {
+      component.setLongName(reportComponent.getPath());
+      if (reportComponent.hasPath()) {
+        component.setPath(reportComponent.getPath());
+      }
+      if (reportComponent.hasLanguage()) {
+        component.setLanguage(reportComponent.getLanguage());
+      }
+    }
+    if (parentModule != null) {
+      component.setParentProjectId(parentModule.getId());
+      component.setProjectUuid(parentModule.projectUuid());
+      component.setModuleUuid(parentModule.uuid());
+      component.setModuleUuidPath(reportComponent.getType().equals(Constants.ComponentType.MODULE) ?
+        parentModule.moduleUuidPath() + component.uuid() + ComponentDto.MODULE_UUID_PATH_SEP :
+        parentModule.moduleUuidPath());
+    } else {
+      component.setProjectUuid(uuid);
+      component.setModuleUuidPath(ComponentDto.MODULE_UUID_PATH_SEP + component.uuid() + ComponentDto.MODULE_UUID_PATH_SEP);
+    }
+    return component;
+  }
+
+  private boolean updateComponent(ComponentDto existingComponent, ComponentDto newComponent) {
+    boolean isUpdated = false;
+    if (Scopes.PROJECT.equals(existingComponent.scope())) {
+      if (!newComponent.name().equals(existingComponent.name())) {
+        existingComponent.setName(newComponent.name());
+        isUpdated = true;
+      }
+      if (!StringUtils.equals(existingComponent.description(), newComponent.description())) {
+        existingComponent.setDescription(newComponent.description());
+        isUpdated = true;
+      }
+    }
+
+    if (!StringUtils.equals(existingComponent.moduleUuid(), newComponent.moduleUuid())) {
+      existingComponent.setModuleUuid(newComponent.moduleUuid());
+      isUpdated = true;
+    }
+    if (!existingComponent.moduleUuidPath().equals(newComponent.moduleUuidPath())) {
```

Comment:
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'moduleUuidPath' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Possible null pointer dereference in org.sonar.server.computation.step.PersistComponentsStep.updateComponent(ComponentDto, ComponentDto) due to return value of called method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)